README.md
# The Knowledge Graph notebook

**_LinkedDataHub_ (LDH) is open source software you can use to manage data, create visualizations and build apps on RDF Knowledge Graphs.**

What's new in LinkedDataHub v3? Watch this video for a feature overview:
[![What's new in LinkedDataHub v3? Feature overview](https://img.youtube.com/vi/phRL6QtVTG0/0.jpg)](https://www.youtube.com/watch?v=phRL6QtVTG0)

We started the project with the intention to use it for Linked Data publishing, but gradually realized that we've built a multi-purpose data-driven platform.

We are building LinkedDataHub primarily for:
* researchers who need an RDF-native notebook that can consume and collect Linked Data and SPARQL documents and follows the [FAIR principles](https://www.go-fair.org/fair-principles/)
* developers who are looking for a declarative full stack framework for Knowledge Graph application development, with out-of-the-box UI and API

What makes LinkedDataHub unique is its completely _data-driven architecture_: applications and documents are defined as data, managed using a single generic HTTP API and presented using declarative technologies. The default application structure and user interface are provided, but they can be completely overridden and customized. Unless a custom server-side processing is required, no imperative code such as Java or JavaScript needs to be involved at all.

**Follow the [Get started](https://atomgraph.github.io/LinkedDataHub/linkeddatahub/docs/get-started/) guide to LinkedDataHub.** The setup and basic configuration sections are provided below and should get you running.

## Setup

<details>
  <summary>Click to expand</summary>

### Prerequisites

* `bash` shell 4.x. It should be included by default on Linux. On Windows you can install the [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install-win10).
* Java's [`keytool`](https://docs.oracle.com/en/java/javase/11/tools/keytool.html) available on `$PATH`. It comes with the JDK.
* [`openssl`](https://www.openssl.org/) available on `$PATH`
* [Docker](https://docs.docker.com/install/) installed. At least 8GB of memory dedicated to Docker is recommended.
* [Docker Compose](https://docs.docker.com/compose/install/) installed

### Steps

  1. [Fork](https://guides.github.com/activities/forking/) this repository and clone the fork into a folder
  2. In the folder, create an `.env` file and fill out the missing values (you can use [`.env_sample`](https://github.com/AtomGraph/LinkedDataHub/blob/master/.env_sample) as a template). For example:
     ```
     COMPOSE_CONVERT_WINDOWS_PATHS=1
     COMPOSE_PROJECT_NAME=linkeddatahub
     
     PROTOCOL=https
     PROXY_HTTP_PORT=81
     PROXY_HTTPS_PORT=4443
     HOST=localhost
     ABS_PATH=/
     
     OWNER_MBOX=john@doe.com
     OWNER_GIVEN_NAME=John
     OWNER_FAMILY_NAME=Doe
     OWNER_ORG_UNIT=My unit
     OWNER_ORGANIZATION=My org
     OWNER_LOCALITY=Copenhagen
     OWNER_STATE_OR_PROVINCE=Denmark
     OWNER_COUNTRY_NAME=DK
     ```
  3. Setup SSL certificates/keys by running this from command line (replace `$owner_cert_pwd` and `$secretary_cert_pwd` with your own passwords):
     ```
     ./scripts/setup.sh .env ssl $owner_cert_pwd $secretary_cert_pwd 3650
     ```
     The script will create an `ssl` sub-folder where the SSL certificates and/or public keys will be placed.
  4. Launch the application services by running this from command line:
     ```
     docker-compose up --build
     ```
     It will build LinkedDataHub's Docker image, start its container and mount the following sub-folders:
     - `data` where the triplestore(s) will persist RDF data
     - `uploads` where LDH stores content-hashed file uploads
     The first should take around half a minute as datasets are being loaded into triplestores. After a successful startup, the last line of the Docker log should read something like:
     ```
     linkeddatahub_1     | 09-Feb-2021 14:18:10.536 INFO [main] org.apache.catalina.startup.Catalina.start Server startup in [32609] milliseconds
     ```
  5. Install `ssl/owner/keystore.p12` into a web browser of your choice (password is the `$owner_cert_pwd` value supplied to `setup.sh`)
     - Google Chrome: `Settings > Advanced > Manage Certificates > Import...`
     - Mozilla Firefox: `Options > Privacy > Security > View Certificates... > Import...`
     - Apple Safari: The file is installed directly into the operating system. Open the file and import it using the [Keychain Access](https://support.apple.com/guide/keychain-access/what-is-keychain-access-kyca1083/mac) tool (drag it to the `local` section).
     - Microsoft Edge: Does not support certificate management, you need to install the file into Windows. [Read more here](https://social.technet.microsoft.com/Forums/en-US/18301fff-0467-4e41-8dee-4e44823ed5bf/microsoft-edge-browser-and-ssl-certificates?forum=win10itprogeneral).
  6. Open **https://localhost:4443/** in that web browser

  ### Notes

  * You will likely get a browser warning such as `Your connection is not private` in Chrome or `Warning: Potential Security Risk Ahead` in Firefox due to the self-signed server certificate. Ignore it: click `Advanced` and `Proceed` or `Accept the risk` to proceed.
    * If this option does not appear in Chrome (as observed on some MacOS), you can open `chrome://flags/#allow-insecure-localhost`, switch `Allow invalid certificates for resources loaded from localhost` to `Enabled` and restart Chrome
  * `.env_sample` and `.env` files might be invisible in MacOS Finder which hides filenames starting with a dot. You should be able to [create it using Terminal](https://stackoverflow.com/questions/5891365/mac-os-x-doesnt-allow-to-name-files-starting-with-a-dot-how-do-i-name-the-hta) however.
  * On Linux your user may need to be a member of the `docker` group. Add it using
  ```
  sudo usermod -aG docker ${USER}
  ```
  and re-login with your user. An alternative, but not recommended, is to run
  ```
  sudo docker-compose up
  ```
</details>

## Configuration

<details>
  <summary>Click to expand</summary>

  ### Base URI

  A common case is changing the base URI from the default `https://localhost:4443/` to your own.

  Lets use `https://ec2-54-235-229-141.compute-1.amazonaws.com/linkeddatahub/` as an example. We need to split the URI into components and set them in the `.env` file using the following parameters:
  ```
  PROTOCOL=https
  HTTP_PORT=80
  HTTPS_PORT=443
  HOST=ec2-54-235-229-141.compute-1.amazonaws.com
  ABS_PATH=/linkeddatahub/
  ```

  `ABS_PATH` is required, even if it's just `/`.

  ### Dataspaces

  Dataspaces are configured in [`config/system-varnish.trig`](https://github.com/AtomGraph/LinkedDataHub/blob/master/config/system-varnish.trig). Relative URIs will be resolved against the base URI configured in the `.env` file.

_:warning: Do not use blank nodes to identify applications or services. We recommend using the `urn:` URI scheme, since LinkedDataHub application resources are not accessible under their own dataspace._

  ### Environment

  LinkedDataHub supports a range of configuration options that can be passed as environment parameters in `docker-compose.yml`. The most common ones are:

  <dl>
    <dt><code>CATALINA_OPTS</code></dt>
    <dd>Tomcat's <a href="https://tomcat.apache.org/tomcat-9.0-doc/RUNNING.txt">command line options</a></dd>
    <dt><code>SELF_SIGNED_CERT</code></dt>
    <dd><code>true</code> if the server certificate is self-signed</dd>
    <dt><code>SIGN_UP_CERT_VALIDITY</code></dt>
    <dd>Validity of the WebID certificates of signed up users (<em>not the owner's</em>)</dd>
    <dt><code>IMPORT_KEEPALIVE</code></dt>
    <dd>The period for which the data import can keep an open HTTP connection before it times out, in ms. The larger files are being imported, the longer it has to be in order for the import to complete.</dd>
    <dt><code>MAX_CONTENT_LENGTH</code></dt>
    <dd>Maximum allowed size of the request body, in bytes</dd>
    <dt><code>MAIL_SMTP_HOST</code></dt>
    <dd>Hostname of the mail server</dd>
    <dt><code>MAIL_SMTP_PORT</code></dt>
    <dd>Port number of the mail server</dd>
    <dt><code>GOOGLE_CLIENT_ID</code></dt>
    <dd>OAuth 2.0 Client ID from Google. When provided, enables the <samp>Login with Google</samp> authentication method.</dd>
    <dt><code>GOOGLE_CLIENT_SECRET</code></dt>
    <dd>Client secret from Google</dd>
  </dl>

  ## Reset

  If you need to start fresh and wipe the existing setup (e.g. after configuring a new base URI), you can do that using
  ```
  sudo rm -rf data uploads && docker-compose down -v
  ```

_:warning: This will **remove the persisted data and files** as well as Docker volumes._
</details>

## [Documentation](https://atomgraph.github.io/LinkedDataHub/linkeddatahub/docs/)

* [Get started](https://atomgraph.github.io/LinkedDataHub/linkeddatahub/docs/get-started/)
* [Reference](https://atomgraph.github.io/LinkedDataHub/linkeddatahub/docs/reference/)
* [User guide](https://atomgraph.github.io/LinkedDataHub/linkeddatahub/docs/user-guide/)

## [Command line interface](https://atomgraph.github.io/LinkedDataHub/linkeddatahub/docs/reference/command-line-interface/)

LinkedDataHub CLI wraps the HTTP API into a set of shell scripts with convenient parameters. The scripts can be used for testing, automation, scheduled execution and such. It is usually much quicker to perform actions using CLI rather than the user interface, as well as easier to reproduce.

The scripts can be found in the [`scripts`](https://github.com/AtomGraph/LinkedDataHub/tree/master/scripts) subfolder.

_:warning: The CLI scripts internally use [Jena's CLI commands](https://jena.apache.org/documentation/tools/). Set up the Jena environment before running the scripts._

An environment variable `JENA_HOME` is used by all the command line tools to configure the class path automatically for you. You can set this up as follows:

**On Linux / Mac**

    export JENA_HOME=the directory you downloaded Jena to
    export PATH="$PATH:$JENA_HOME/bin"

**Get the [source code](https://github.com/AtomGraph/LinkedDataHub-Apps)**

_:warning: Before running app installation scripts that use LinkedDataHub's CLI scripts, set the `SCRIPT_ROOT` environmental variable to the [`scripts`](https://github.com/AtomGraph/LinkedDataHub/tree/master/scripts) subfolder of your LinkedDataHub fork or clone._ For example:

    export SCRIPT_ROOT="/c/Users/namedgraph/WebRoot/AtomGraph/LinkedDataHub/scripts"

## How to get involved

* contribute a new LDH application or modify [one of ours](https://github.com/AtomGraph/LinkedDataHub-Apps)
* work on [good first issues](../../contribute)
* work on the features in our [Roadmap](../../wiki/Roadmap)
* join our [community](#community)

## Test suite

LinkedDataHub includes an HTTP [test suite](https://github.com/AtomGraph/LinkedDataHub/tree/master/http-tests). The server implementation is also covered by the [Processor test suite](https://github.com/AtomGraph/Processor/tree/master/http-tests).

![HTTP-tests](https://github.com/AtomGraph/LinkedDataHub/workflows/HTTP-tests/badge.svg?branch=master)
![HTTP-tests](https://github.com/AtomGraph/LinkedDataHub/workflows/HTTP-tests/badge.svg?branch=develop)

## Dependencies

### Browser

* [Saxon-JS](https://www.saxonica.com/saxon-js/)
* [SPARQLMap](https://github.com/AtomGraph/SPARQLMap)
* [Google Maps](https://developers.google.com/maps/)
* [Google Charts](https://developers.google.com/chart)

### Java

* [Jersey](https://eclipse-ee4j.github.io/jersey/)
* [XOM](http://www.xom.nu)
* [JavaMail](https://javaee.github.io/javamail/)
* [Guava](https://github.com/google/guava)
* [java-jwt](https://github.com/auth0/java-jwt)
* [ExpiringMap](https://github.com/jhalterman/expiringmap)
* [CSV2RDF](https://github.com/AtomGraph/CSV2RDF)
* [Processor](https://github.com/AtomGraph/Processor)
* [Web-Client](https://github.com/AtomGraph/Web-Client)

### Docker

* [atomgraph/nginx](https://hub.docker.com/r/atomgraph/nginx)
* [atomgraph/fuseki](https://hub.docker.com/r/atomgraph/fuseki)
* [atomgraph/varnish](https://hub.docker.com/r/atomgraph/varnish)
* [namshi/smtp](https://hub.docker.com/r/namshi/smtp)

## Support

Please [report issues](https://github.com/AtomGraph/LinkedDataHub/issues) if you've encountered a bug or have a feature request.

Commercial consulting, development, and support are available from [AtomGraph](https://atomgraph.com).

## Community

* [linkeddatahub@groups.io](https://groups.io/g/linkeddatahub) (mailing list)
* [@atomgraphhq](https://twitter.com/atomgraphhq) on Twitter
* [AtomGraph](https://www.linkedin.com/company/atomgraph/) on LinkedIn
* W3C [Declarative Linked Data Apps Community Group](http://www.w3.org/community/declarative-apps/)

CHANGELOG
## [3.0.6] - 2022-03-14
### Added
- Interactivity to the graph SVG layout
- Notifications to the requesting agent when its access request is granted (requires email server)
- JSON-LD export option for documents
- `append-content` CLI script that appends content resources to document
- `create-file` CLI script automatically recognizes the MIME type of the file being uploaded
- Linked Data browser functionality in graph layout mode
- Javadoc comments

### Changed
- The setup script requirements relaxed to make it easier to run on MacOS
- Fixed CSV imports
- Linked Data browser now supports relative URIs
- Upgraded Java from 11 to 17
- A built-in HTTP API constraint does not allow `PUT` on documents without the document description in request body
- A built-in HTTP API constraint does not allow to `DELETE` the root document
- A built-in HTTP API constraint does not allow to `DELETE` or `PUT` the the app owner's and secretary's WebID documents
- Shell script interpreter line

### Removed
- Dydra-specific code

## [3.0.3] - 2022-02-16
### Added
- Ability to copy (fork) RDF data into the local dataspace
- Block-based content layout (`ldh:ContentList` mode) and editor
- Login with Google (OpenID Connect)
- Ability to load JSON-LD data from `<script>` elements in HTML
- Namespace endpoint, which is an in-memory SPARQL endpoint over the app's ontology

### Changed
- The HTTP CRUD API is now Graph Store Protocol, not Linked Data Templates
- HTML documents are hydrated HTML fragments over AJAX
- XSLT stylesheets now load constraints and constructors using SPARQL over the namespace endpoint
- The URIs of ontology terms are not relative to the app's base URI anymore
- Additional assertions added to external ontology terms instead of subclassing them
- Every UI state generates a distinct URL which is loaded consistently on both server- and client-side
- Upgraded Jena to 4.3.2
- Upgraded Saxon-JS to 2.3
- Upgraded Fuseki, Varnish and nginx Docker images

### Removed
- Linked Data Templates support (still supported by Processor)

## [2.2.9] - 2021-04-22
### Added
- HTTP smoke tests for SPARQL endpoint and Graph Store
- HTTP test for RDF import without mapping query
- `add-data.sh` CLI script which POSTs RDF data to URL
- `ExceptionMapper` constructors with injection in order to align with Processor
- An option to override request URI using the `?uri=` URL param, implemented in `ApplicationFilter`
- `Dispatcher` as the new "entrypoint" JAX-RS resource which routes between `ResourceBase` (if app is not empty) and `ProxyResourceBase` (if app is empty)
- Missing XML namespace definitions to client-side XSLT stylesheets
- `$output-json-ld` parameter in `xhtml:Script` template which outputs the RDF document as JSON-LD in the `<script>` element

### Changed
- `select-labelled` query in the end-user dataset to include a default graph pattern
- `spin:query` property is now optional for `apl:RDFImport`
- Entrypoint script logic to load agent metadata only when `$LOAD_DATASETS` is true
- Injecting `Optional<Application>`, `Optional<Service>`, `Optional<Ontology>` instead of `Application`, `Service`, `Ontology`
- Using `javax.inject.Provider<>` for injection into providers that are not in the request scope
- If no application matches request URI, `NotFoundException` is not thrown anymore -- `Optional.empty()` is used as application instead
- Auth filters skipped if the matched application is not an instance of `lapp:EndUserApplication` or `lapp:AdminApplication`
- Simplified `ResourceBase::describe` by removing the `?uri=` indirection logic

### Removed
- Proxy injections from injection factory binders


## [2.1.55] - 2021-03-26
### Added
- varnish-admin service that proxy-caches the fuseki-admin triplestore
- `purge_backend_cache` function to the HTTP test runner script run.sh
- `purge_backend_cache` calls to clear proxy caches before each HTTP script
- `BackendInvalidationFilter` response filter with backend proxy cache invalidation heuristics
- Basic environment variable documentation to README

### Changed
- Upgraded Processor and Web-Client to the latest versions
- Upgraded Saxon-JS to 2.1
- End-user and admin Services passed to import `Executor` instead of `DatasetAccessor`
- Defined HTTP method -> ACL mode mapping as the `AuthorizationFilter.ACCESS_MODES` map

### Removed
- Unused Docker mounts from linkeddatahub service
- `ban()` calls from `ResourceBase` -- now handled by the `BackendInvalidationFilter`

## [2.1.49] - 2021-03-19
### Added
- `apl:baseUri` as a static XSLT stylesheet param

### Changed
- `bs2:PropertyControl` XSLT mode can handle multiple RDF types on a resource
- Replaced error alerts with inline HTML warning blocks
- Refactored `bs2:SignUp` template to make it more extensible

### Removed
- `ORDER BY` in `apl:ResultCounts` mode in client.xsl


## [2.1.28] - 2021-03-06
### Added
- `$request_base` parameter support in scripts allows to use a base URI for HTTP requests which is different from the RDF dataset base URI. Useful when multiple LDH instances on different domains or port numbers are backed by the same dataset. E.g. one with WebID-TLS auth enabled and the other without.
- Dydra-specific `QuadStoreClient` and `GraphStoreClient` with support for [asynchonous GSP requests](https://api.dydra.com/graphstore/asynchronous.html)

### Changed
- Ontology classes that used to be in the `ns:` namespace (`${base}ns#`) moved to `nsds:` (`${base}ns/domain/system#`)
- Ontology classes that used to be in the `def:` namespace (`${base}def#`) moved to `nsdd:` (`${base}ns/domain/default#`)
- `python` usages replaced with `python2` in CLI scripts
- Mounting only `ssl/owner/public.pem` instead of the whole `ssl/owner` folder which includes the private key

### Removed
- Expensive join with the provenance graph from the `laclt:ConstructAgentItem` query. As a result, `dct:created` value is not included in agent's description.

scripts/setup.sh
#!/usr/bin/env bash
set -e

# check required bash version
[ "${BASH_VERSINFO:-0}" -lt 4 ] && echo "Bash version too old. Bash 4 is required for associative array support." && exit 1

if [ "$#" -ne 5 ]; then
    echo "Usage:   $0" '$env_file $out_folder $owner_cert_pwd $secretary_cert_pwd $validity' >&2
    echo "Example: $0 .env ssl Password Password 3650" >&2
    exit 1
fi

env_file="$1"
out_folder="$2"

printf "### Output folder: %s\n" "$out_folder"

proxy_host="nginx"

server_cert="${out_folder}/server/server.crt"
server_public_key="${out_folder}/server/server.key"

owner_alias="owner"
owner_keystore="${out_folder}/owner/keystore.p12"
owner_cert_pwd="$3"
owner_cert="${out_folder}/owner/cert.pem"
owner_public_key="${out_folder}/owner/public.pem"

secretary_alias="secretary"
secretary_keystore="${out_folder}/secretary/keystore.p12"
secretary_cert="${out_folder}/secretary/cert.pem"
secretary_cert_pwd="$4"
secretary_cert_dname="CN=LDH, OU=LDH, O=AtomGraph, L=Copenhagen, ST=Denmark, C=DK"

validity="$5"

# append secretary cert password to the env_file
echo "" >> "$env_file"
echo "SECRETARY_CERT_PASSWORD=${secretary_cert_pwd}" >> "$env_file"

declare -A env

# read file line by line and populate the array. Field separator is "="
while IFS='=' read -r k v; do
    if [ -n "$k" ] ; then env["$k"]="$v"; fi
done < "$env_file"

if [ -z "${env["PROTOCOL"]}" ]; then
    echo "Configuration is incomplete: PROTOCOL is missing"
    exit 1
fi
if [ -z "${env["HTTPS_PORT"]}" ]; then
    echo "Configuration is incomplete: HTTPS_PORT is missing"
    exit 1
fi
if [ -z "${env["HTTP_PORT"]}" ]; then
    echo "Configuration is incomplete: HTTP_PORT is missing"
    exit 1
fi
if [ -z "${env["HOST"]}" ]; then
    echo "Configuration is incomplete: HOST is missing"
    exit 1
fi
if [ -z "${env["ABS_PATH"]}" ]; then
    echo "Configuration is incomplete: ABS_PATH is missing"
    exit 1
fi

if [ "${env["PROTOCOL"]}" = "https" ]; then
    if [ "${env["HTTPS_PORT"]}" = 443 ]; then
        base_uri="${env["PROTOCOL"]}://${env["HOST"]}${env["ABS_PATH"]}"
    else
        base_uri="${env["PROTOCOL"]}://${env["HOST"]}:${env["HTTPS_PORT"]}${env["ABS_PATH"]}"
    fi
else
    if [ "${env["HTTP_PORT"]}" = 80 ]; then
        base_uri="${env["PROTOCOL"]}://${env["HOST"]}${env["ABS_PATH"]}"
    else
        base_uri="${env["PROTOCOL"]}://${env["HOST"]}:${env["HTTP_PORT"]}${env["ABS_PATH"]}"
    fi
fi

printf "\n### Base URI: %s\n" "$base_uri"

### SERVER CERT ###

mkdir -p "$out_folder"/server

# crude check if the host is an IP address
IP_ADDR_MATCH=$(echo "${env["HOST"]}" | grep -oE "\b([0-9]{1,3}\.){3}[0-9]{1,3}\b" || test $? = 1)

if [ -n "$IP_ADDR_MATCH" ]; then
    if [ -n "$proxy_host" ]; then
        san="subjectAltName=IP:${env["HOST"]},DNS:${proxy_host}" # IP address - special case for localhost
    else
        san="subjectAltName=IP:${env["HOST"]}" # IP address
    fi
else
    if [ -n "$proxy_host" ]; then
        san="subjectAltName=DNS:${env["HOST"]},DNS:${proxy_host}" # hostname - special case for localhost
    else
        san="subjectAltName=DNS:${env["HOST"]}" # hostname
    fi
fi

# openssl <= 1.1.1
openssl req -x509 -newkey rsa:4096 -sha256 -days 3650 -nodes \
  -keyout "$server_public_key" \
  -out "$server_cert" \
  -subj "/CN=${env["HOST"]}/OU=LinkedDataHub/O=AtomGraph/L=Copenhagen/C=DK" \
  -extensions san \
  -config <(echo '[req]'; echo 'distinguished_name=req';
            echo '[san]'; echo "$san")

# openssl >= 1.1.1
#openssl req -x509 -newkey rsa:4096 -sha256 -days 3650 -nodes \
#  -keyout "$server_public_key" \
#  -out "$server_cert" \
#  -subj "/CN=${env["HOST"]}/OU=LinkedDataHub/O=AtomGraph/L=Copenhagen/C=DK" \
#  -addext "$san"

### OWNER CERT ###

if [ -z "${env["OWNER_GIVEN_NAME"]}" ]; then
    echo "Configuration is incomplete: OWNER_GIVEN_NAME is missing"
    exit 1
fi
if [ -z "${env["OWNER_FAMILY_NAME"]}" ]; then
    echo "Configuration is incomplete: OWNER_FAMILY_NAME is missing"
    exit 1
fi
if [ -z "${env["OWNER_ORG_UNIT"]}" ]; then
    echo "Configuration is incomplete: OWNER_ORG_UNIT is missing"
    exit 1
fi
if [ -z "${env["OWNER_ORGANIZATION"]}" ]; then
    echo "Configuration is incomplete: OWNER_ORGANIZATION is missing"
    exit 1
fi
if [ -z "${env["OWNER_LOCALITY"]}" ]; then
    echo "Configuration is incomplete: OWNER_LOCALITY is missing"
    exit 1
fi
if [ -z "${env["OWNER_STATE_OR_PROVINCE"]}" ]; then
    echo "Configuration is incomplete: OWNER_STATE_OR_PROVINCE is missing"
    exit 1
fi
if [ -z "${env["OWNER_COUNTRY_NAME"]}" ]; then
    echo "Configuration is incomplete: OWNER_COUNTRY_NAME is missing"
    exit 1
fi

owner_uuid=$(uuidgen | tr '[:upper:]' '[:lower:]') # lowercase
owner_uri="${base_uri}admin/acl/agents/${owner_uuid}/#this"

printf "\n### Owner's WebID URI: %s\n" "$owner_uri"

owner_cert_dname="CN=${env["OWNER_GIVEN_NAME"]} ${env["OWNER_FAMILY_NAME"]}, OU=${env["OWNER_ORG_UNIT"]}, O=${env["OWNER_ORGANIZATION"]}, L=${env["OWNER_LOCALITY"]}, ST=${env["OWNER_STATE_OR_PROVINCE"]}, C=${env["OWNER_COUNTRY_NAME"]}"
printf "\n### Owner WebID certificate's DName attributes: %s\n" "$owner_cert_dname"

mkdir -p "$out_folder"/owner

keytool \
    -genkeypair \
    -alias "$owner_alias" \
    -keyalg RSA \
    -storetype PKCS12 \
    -keystore "$owner_keystore" \
    -storepass "$owner_cert_pwd" \
    -keypass "$owner_cert_pwd" \
    -dname "$owner_cert_dname" \
    -ext "SAN=uri:${owner_uri}" \
    -validity "$validity"

# convert owner's certificate to PEM

openssl \
    pkcs12 \
    -in "$owner_keystore" \
    -passin pass:"$owner_cert_pwd" \
    -out "$owner_cert" \
    -passout pass:"$owner_cert_pwd"

# convert owner's public key to PEM

openssl \
    pkcs12 \
    -in "$owner_keystore" \
    -passin pass:"$owner_cert_pwd" \
    -nokeys \
    -out "$owner_public_key"

### SECRETARY CERT ###

mkdir -p "$out_folder"/secretary

secretary_uuid=$(uuidgen | tr '[:upper:]' '[:lower:]') # lowercase
secretary_uri="${base_uri}admin/acl/agents/${secretary_uuid}/#this"

printf "\n### Secretary's WebID URI: %s\n" "$secretary_uri"

printf "\n### Secretary WebID certificate's DName attributes: %s\n" "$secretary_cert_dname"

keytool \
    -genkeypair \
    -alias "$secretary_alias" \
    -keyalg RSA \
    -storetype PKCS12 \
    -keystore "$secretary_keystore" \
    -storepass "$secretary_cert_pwd" \
    -keypass "$secretary_cert_pwd" \
    -dname "$secretary_cert_dname" \
    -ext "SAN=uri:${secretary_uri}" \
    -validity "$validity"

# convert secretary's certificate to PEM

openssl \
    pkcs12 \
    -in "$secretary_keystore" \
    -passin pass:"$secretary_cert_pwd" \
    -out "$secretary_cert" \
    -passout pass:"$secretary_cert_pwd"